### ISSUE TITLE
Feature Request: Commit Filtering

### ISSUE THREAD

**Issue by therealbnut** [buildkite/feedback#130]
It would be great to have commit filtering, exactly like **Branch Filtering** in a build step's settings, except based on the commit message. This would allow us to do things like `[skip ci]` except for things like `[skip unit tests]` or `[xcode-7.1]` etc.

**Comment by keithpitt** [comment]
Yeah! We've thought about this sort of thing in the past, just haven't gotten around to looking into it more. There is a way you can implement this yourself for the time being though!

If you switch to using our `pipeline.yml` stuff instead of configuring the pipeline in the UI: https://buildkite.com/docs/agent/uploading-pipelines - it'll give you much more control about how the pipeline works. You can even have dynamic pipelines (so instead of being a static file, you can generate the pipeline using code and feed it to us).

(Below is some hacky, incorrect ruby code, but gives you an idea of how to do it)

pipeline.rb
```ruby
STEPS = [
  { name: "Run the unit tests!", tag: "unit tests", command: "..." },
  { name: "Xcode party!", tag: "xcode-7.1", command: "..." },
]

message = ENV["BUILDKITE_MESSAGE"]

pipeline = STEPS.inject([]) do |arr, step|
  if message !=~ /[skip #{step.tag}]/
    arr << step
  end
  arr
end

puts JSON.pretty_generate(pipeline)
```

Then as the pipeline upload step:

`ruby pipeline.rb | buildkite-agent pipeline upload`

So you define steps in the ruby file, only output the steps that should _actually_ be run, and pipe the result to the `buildkite-agent pipeline upload` command (which accepts JSON because it's a bit easier to work with when giving it a custom pipeline).

Hope this helps!

**Comment by therealbnut** [comment]
Thanks for the excellent response! We have been looking for an excuse to play with the new pipeline stuff.

A few questions:
 * Is `BUILDKITE_MESSAGE` the commit message, and is it supported? I don't think it's in the [documentation](https://buildkite.com/docs/guides/environment-variables).
 * Does this mean that for each build (commit) you're uploading a new global pipeline, or does this only affect the current build? I don't imagine it would work per-commit for concurrent builds unless it was restricted to the current build only.

**Comment by keithpitt** [comment]
Ah, doh! I can't believe `BUILDKITE_MESSAGE` was missing from the docs. I've fixed that pushed it to production. Yeah, it's officially supported :)

So `buildkite-agent pipeline upload` only affects the current build - which means each pipeline could have it's own unique pipeline. This post has a few screenshots that may give you a better idea of how it all works: http://danthorpe.me/posts/uploading-build-pipelines.html

**Comment by therealbnut** [comment]
Nice! Thank you :)

I'll leave this issue open until we explore doing it with pipelines a bit more.

Speaking of branch filtering, I'd expect something like: `!*-do-the-thing` to match all branches not ending with `-do-the-thing`, but it doesn't seem to be working for us. Perhaps a bug? It seems that `*` and `!` seem to work independently but not together.

**Comment by keithpitt** [comment]
Yeah, you're right! Looks like we didn't support wildcards at the start of the branch config. I've just updated our support so now we support `x` number of wildcards anywhere in the branch config. So you're example should work this time!

**Comment by therealbnut** [comment]
Awesome, thanks!
